File: 1234. Replace the Substring for Balanced String.py
class Solution:
    def balancedString(self, s: str) -> int:
        count = collections.defaultdict(int)
        extra = collections.defaultdict(int)
        for c in s:
            count[c] += 1
        for c in "QWER":
            extra[c] = max(0, count[c]-len(s)//4)
            count[c] = 0
        
        ans = len(s)
        l = r = 0
        while (r < len(s)):
            count[s[r]] += 1
            while l<=r and count[s[l]] > extra[s[l]]:
                count[s[l]] -= 1
                l += 1
                
            if all(count[c]>=extra[c] for c in "QWER"):
                ans = min(ans, r-l+1)
            r += 1
        return ans
        

File: 1234. Replace the Substring for Balanced String.cpp
class Solution {
public:
    int balancedString(string s) {
        vector<int> count(128), extra(128);
        for (auto c : s) ++count[c];
        for (auto c : "QWER") {
            extra[c] = max<int>(0, count[c] - (int)s.size()/4);
            count[c] = 0;
        }
        
        int ans = s.size();
        for (int l=0,r=0; r<s.size(); ++r) {
            ++count[s[r]];
            while (l<=r && count[s[l]]>extra[s[l]]) 
                --count[s[l++]];
            if (count['Q']>=extra['Q'] && count['W']>=extra['W'] && count['E']>=extra['E'] && count['R']>=extra['R'])
                ans = min(ans, r-l+1);
        }
        return ans;
        
    }
};